jans-cedarling/cedarling/src/jwt/decoding_strategy.rs

```diff
@@ -5,140 +5,197 @@
     aud: Option<impl ToString>,
-    req_sub: bool,
-    supported_algs: &[jwt::Algorithm],
-    key_service: &Arc<dyn GetKey>,
+    sub: Option<impl ToString>,
+    validate_nbf: bool,
+    validate_exp: bool,
+    supported_algs: &Vec<jwt::Algorithm>,
+    key_service: &KeyService,
 ) -> Result<T, Error> {
-    let header = jwt::decode_header(jwt).map_err(Error::Parsing)?;
+    let header = jwt::decode_header(jwt).map_err(Error::ParsingError)?;
 
     // reject unsupported algorithms early
     if !supported_algs.contains(&header.alg) {
         return Err(Error::TokenSignedWithUnsupportedAlgorithm(header.alg));
     }
 
-    // Set validator configs
+    // set up validation rules
     let mut validator = jwt::Validation::new(header.alg);
-    validator.validate_nbf = true;
+    validator.required_spec_claims.clear();
+    validator.validate_nbf = validate_nbf;
+    validator.validate_exp = validate_exp;
     if let Some(iss) = iss {
-        validator.set_issuer(&[iss]);
+        validator.set_issuer(&vec![iss]);
+    } else {
+        validator.iss = None;
     }
     if let Some(aud) = aud {
         validator.set_audience(&[aud]);
     } else {
         validator.validate_aud = false;
     }
-    if req_sub {
-        validator.set_required_spec_claims(&["sub"]);
-    }
+    validator.sub = match sub {
+        Some(sub) => Some(sub.to_string()),
+        None => None,
+    };
 
     // fetch decoding key from the KeyService
     let kid = &header
         .kid
         .ok_or_else(|| Error::MissingRequiredHeader("kid".into()))?;
-    let key = key_service.get_key(kid)?;
+    let key = key_service.get_key(kid).map_err(Error::KeyServiceError)?;
     // TODO: handle tokens without a `kid` in the header
 
-    // extract claims
-    let claims = jwt::decode::<T>(jwt, key, &validator)
-        .map_err(Error::Validation)?
+    // decode and validate the jwt
+    let claims = jwt::decode::<T>(&jwt, &key, &validator)
+        .map_err(|e| Error::ValidationError(e.to_string()))?
         .claims;
     Ok(claims)
 }
 
-#[derive(thiserror::Error, Debug)]
-pub enum ParseAlgorithmError {
-    /// Config contains an unimplemented algorithm
-    #[error("algorithim is not yet implemented: {0}")]
-    UnimplementedAlgorithm(String),
-}
-
 /// Converts a string representation of an algorithm to a `jwt::Algorithm` enum.
 ///
-/// This function attempts to map a string representing an algorithm (e.g., "HS256")
-/// to its corresponding `jwt::Algorithm` enum. If the algorithm is unsupported or
-/// unrecognized, an error is returned.
-pub(crate) fn string_to_alg(algorithm: &str) -> Result<jwt::Algorithm, ParseAlgorithmError> {
+/// This function maps algorithm names (e.g., "HS256", "RS256") to corresponding
+/// `jwt::Algorithm` enum values. Returns an error if the algorithm is unsupported.
+///
+/// # Errors
+/// Returns an error if the algorithm is not implemented.
+pub fn string_to_alg(algorithm: &str) -> Result<jwt::Algorithm, Error> {
```

Comment:
why this function return all possible errors in the module (that will newer happen). If it called only when parse Bootsrap config???

Reply:
i'll fix it in #9945 if that's okay.

Reply:
resolved in [c8c8945](https://github.com/JanssenProject/jans/pull/9939/commits/c8c8945883b97b1041fccbba0c10f3c9b17c4d19).